### PR TITLE
Completely refactors Karma

### PR DESCRIPTION
**Pull request by AffectedArc07**
## What Does This PR Do
This PR completely rewrites the karma system to be a lot more sane.

In its current form, Karma is species & jobs, and only those things. It is incredibly flawed and the UI is not dynamic in the slightest, you have to manually update the display every time you add packages, not to mention how non-OOP it is with refunds.

In short, this PR:
- Converts the karma UI to TGUI
    - Yes I know I said no OOC UIs but this is hardly a game-dependant OOC UI, and being able to make it dynamic is enough of a positive over the current form
- The database schema now has a standalone `purchase` column which can have any form of purchase in it, which allows karma to be expanded into other means
    - This also means karma jobs can be completely nuked with no further changes needed
- Karma refunds are now fully automatic. If a package is marked as refundable, the refund will be automatically be applied

Now for the bigger explanation.

This is the current schema
image
image

Its. Disgusting. ITS HORRIBLE. Not only does it break all forms of DB normalisation, its limited to just jobs and species, not to mention data corruption as seen here.

The new schema is much more sane
image

The karma shop also has a new coat of paint. These tabs will automatically be generated based on the defined packages, and the buttons now mean different things.
image

- Blue = purchasable
- Green = already bought
- Grey = cannot afford (with tooltip)

Full screenshot
image

Adding a new karma purchase is now extremely easy to do
```dm
/datum/karma_package/job/blueshield
	database_id = KARMAPACKAGE_JOB_BLUESHIELD
	friendly_name = "Blueshield"
	karma_cost = 30
```

Simply define a package above, and it will be added to the respective category and have its DB rows be fine. You can then do `client.karmaholder.hasPackage(KARMAPACKAGE_JOB_BLUESHIELD)` to see if someone has the purchase or not. Its that easy.

To make a package refundable, you simply add the refundable var.
```dm
/datum/karma_package/job/brigphys
	database_id = KARMAPACKAGE_JOB_BRIGPHYSICIAN
	friendly_name = "Brig Physician"
	karma_cost = 5
	refundable = TRUE
```

This will make the refund automatically apply to the user when they next login, removing the need for them to manually do it.
image

This entire system is much more versatile, much more OOP and modernly programmed, and makes you want to tear your eyes out less.

## Why It's Good For The Game
See my paragraphs above

## Changelog
:cl: AffectedArc07
add: Completely refactored karma
add: Karma refunds now apply automatically if you have them available
add: Added the framework to allow for more things other than jobs/species in the karma store in the future (NOTE: None have been added as of now)
/:cl:

### CODE REVIEW

File: code/modules/client/client_procs.dm [hunk 1]
```diff
@@ -397,6 +397,9 @@
 	if(_2fa_alert)
 		to_chat(src,"<span class='boldannounce'><big>You do not have 2FA enabled. Admin verbs will be unavailable until you have enabled 2FA.</big></span>") // Very fucking obvious
 
+	// This happens asyncronously
+	karmaholder.processRefunds(src.mob)
```
Comment (AffectedArc07) [inline code/modules/client/client_procs.dm:401]:
```suggestion
	karmaholder.processRefunds(mob)
```

File: SQL/updates/31-32.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+# Listen up.
+# In order to run this script on Windows, you need to make sure you have Python **3** installed. Tested on 3.8.2
+# It won't work on 2.7 at all.
+
+# To run this, supply the following args in a command shell
+# python 31-32.py address username password database feedback round
```
Comment (AffectedArc07) [inline SQL/updates/31-32.py:6]:
```suggestion
# python 31-32.py address username password database
```
oops

File: SQL/updates/30-31.sql [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+RENAME TABLE `karma` TO `karma_log`;
+RENAME TABLE `karmatotals` TO `karma_totals`;
+
+# Create new table to track new purchases
+CREATE TABLE `karma_purchases` (
+	`id` INT(11) NOT NULL AUTO_INCREMENT,
+	`ckey` VARCHAR(32) NOT NULL COLLATE 'utf8_general_ci',
+	`purchase` VARCHAR(64) NOT NULL COLLATE 'utf8_general_ci',
+	`purchase_time` DATETIME NOT NULL DEFAULT current_timestamp(),
+	PRIMARY KEY (`id`) USING BTREE,
```
Comment (FlattestGuitar) [inline SQL/updates/30-31.sql:14]: InnoDB only uses b-tree indexes. Specifying index type is redundant. Both here and in the already existing schema. 

Might as well leave this in so it looks prettier I guess?

File: code/modules/karma/karma_packages.dm [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+// If you are ever removing something from the karma store, **DO NOT** delete its definition datum.
+// Just set redunable = TRUE on it so people can get their karma refund
+
+// Jobs
+/datum/karma_package/job
+	category = "Karma Jobs"
+
+/datum/karma_package/job/blueshield
+	database_id = KARMAPACKAGE_JOB_BLUESHIELD
```
Comment (FlattestGuitar) [inline code/modules/karma/karma_packages.dm:30]: I wouldn't rely on defines to keep these names consistent, but this is minor.

File: code/modules/karma/karma_packages.dm [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	var/karma_cost = 0
+	/// Is this package refundable?
+	var/refundable = FALSE
+	/// Category to show in under the karma store
+	var/category = "UNSET"
+	/// List of "Meta" packages that arent entries, just organisers
+	var/list/meta_packages = list(/datum/karma_package/job, /datum/karma_package/species)
+
+/datum/karma_package/vv_edit_var(var_name, var_value)
+	return FALSE // fuck off
```
[the remark this answers is not in the document]
Comment (AffectedArc07) [reply code/modules/karma/karma_packages.dm:19]: Yes I do actually, we're a close family.

File: code/modules/karma/karma_packages.dm (hunk 4 above)
Comment (AffectedArc07) [reply code/modules/karma/karma_packages.dm:30]: Its safer than trusting magic strings

File: SQL/updates/30-31.sql (hunk 3 above)
Comment (AffectedArc07) [reply SQL/updates/30-31.sql:14]: I would rather keep consistency, and this is what shows as the create code for the table.
image

*obviously the increment offset isnt included*

**Review by dearmochi — APPROVED** [review]

**Pull request merged**